core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DocumentVersionInfo.java

```diff
@@ -19,57 +19,181 @@
 import static com.mongodb.stitch.core.services.mongodb.remote.sync.internal.DataSynchronizer.DOCUMENT_VERSION_FIELD;
 
 import com.mongodb.stitch.core.internal.common.BsonUtils;
+
+import java.util.UUID;
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import org.bson.BsonBoolean;
 import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonInt64;
+import org.bson.BsonString;
 import org.bson.BsonValue;
 
+
 final class DocumentVersionInfo {
-  private final BsonValue version;
-  private final BsonDocument filter;
+  @Nullable private final VersionValues versionValues;
+  @Nullable private final BsonDocument versionDoc;
+  @Nullable private final BsonDocument filter;
+
+  private class VersionValues {
+    final int syncProtocolVersion;
+    final String instanceId;
+    final long versionCounter;
+
+    VersionValues(
+            final int syncProtocolVersion,
+            final String instanceId,
+            final long versionCounter) {
+      this.syncProtocolVersion = syncProtocolVersion;
+      this.instanceId = instanceId;
+      this.versionCounter = versionCounter;
+    }
+  }
+
+  private static final String SYNC_PROTOCOL_VERSION_FIELD = "spv";
+  private static final String INSTANCE_ID_FIELD = "id";
+  private static final String VERSION_COUNTER_FIELD = "v";
+
+  private DocumentVersionInfo(
+          @Nullable final BsonDocument version,
+          @Nullable final BsonValue documentId
+  ) {
+    if (version != null) {
+      this.versionDoc = version;
+      this.versionValues = new VersionValues(
```

Comment:
[minor][related to comment below] We should be checking that these keys exist on the `versionDoc` before reading them.

If it the doc is null _or_ they do not exist _and_ this is a disallowed state, we should throw an `IllegalStateException`.

Reply:
From what I can tell here, if the `versionDoc` is null, that _should_ be an OK state, and is covered by `isNonEmptyVersion`. Maybe the exception should be if the versionDoc is malformed? Just food for thought– I think to an extent, there's a stylistic choice here.

Reply:
Definitely agree that this class is a little funky, I refactored it a little bit per your comments and eric's comments to make it harder to reach an illegal state.